In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import io
import requests
url = "https://api.covidtracking.com/v1/states/daily.csv" # Access date and time : Feb 8, 2021, 10.30 A.M.
s=requests.get(url).content

In [3]:
df = pd.read_csv(io.StringIO(s.decode('utf-8')))

In [4]:
df.to_csv('../Dataset/dataset.csv', index=False, header=True) # Saving the dataset for reuse in future time

In [4]:
pd.set_option('display.max_columns', None) # To display all columns
df.head()

date state  positive  probableCases   negative  pending  \
0  20210209    AK   53809.0            NaN        NaN      NaN   
1  20210209    AL  474666.0       102040.0  1825185.0      NaN   
2  20210209    AR  308848.0        63615.0  2297201.0      NaN   
3  20210209    AS       0.0            NaN     2140.0      NaN   
4  20210209    AZ  787268.0        52032.0  2834530.0      NaN   

  totalTestResultsSource  totalTestResults  hospitalizedCurrently  \
0        totalTestsViral         1561249.0                   35.0   
1  totalTestsPeopleViral         2197811.0                 1441.0   
2        totalTestsViral         2542434.0                  775.0   
3        totalTestsViral            2140.0                    NaN   
4        totalTestsViral         7037126.0                 2744.0   

   hospitalizedCumulative  inIcuCurrently  inIcuCumulative  \
0                  1223.0             NaN              NaN   
1                 43499.0             NaN           2583.0   
2                 14174.0           276.0              NaN   
3                     NaN             NaN              NaN   
4                 54967.0           797.0              NaN   

   onVentilatorCurrently  onVentilatorCumulative  recovered dataQualityGrade  \
0                    6.0                     NaN        NaN                A   
1                    NaN                  1463.0   252880.0                A   
2                  137.0                  1464.0   288774.0               A+   
3                    NaN                     NaN        NaN              NaN   
4                  524.0                     NaN   108965.0               A+   

      lastUpdateEt          dateModified  checkTimeEt    death  hospitalized  \
0   2/9/2021 03:59  2021-02-09T03:59:00Z  02/08 22:59    280.0        1223.0   
1   2/9/2021 11:00  2021-02-09T11:00:00Z  02/09 06:00   8579.0       43499.0   
2   2/9/2021 00:00  2021-02-09T00:00:00Z  02/08 19:00   5148.0       14174.0   
3  12/1/2020 00:00  2020-12-01T00:00:00Z  11/30 19:00      0.0           NaN   
4   2/9/2021 00:00  2021-02-09T00:00:00Z  02/08 19:00  14286.0       54967.0   

            dateChecked  totalTestsViral  positiveTestsViral  \
0  2021-02-09T03:59:00Z        1561249.0             65100.0   
1  2021-02-09T11:00:00Z              NaN                 NaN   
2  2021-02-09T00:00:00Z        2542434.0                 NaN   
3  2020-12-01T00:00:00Z           2140.0                 NaN   
4  2021-02-09T00:00:00Z        7037126.0                 NaN   

   negativeTestsViral  positiveCasesViral  deathConfirmed  deathProbable  \
0           1494355.0                 NaN             NaN            NaN   
1                 NaN            372626.0          6783.0         1796.0   
2           2297201.0            245233.0          4119.0         1029.0   
3                 NaN                 0.0             NaN            NaN   
4                 NaN            735236.0         12702.0         1584.0   

   totalTestEncountersViral  totalTestsPeopleViral  totalTestsAntibody  \
0                       NaN                    NaN                 NaN   
1                       NaN              2197811.0                 NaN   
2                       NaN                    NaN                 NaN   
3                       NaN                    NaN                 NaN   
4                       NaN              3569766.0            424813.0   

   positiveTestsAntibody  negativeTestsAntibody  totalTestsPeopleAntibody  \
0                    NaN                    NaN                       NaN   
1                    NaN                    NaN                  109997.0   
2                    NaN                    NaN                       NaN   
3                    NaN                    NaN                       NaN   
4                    NaN                    NaN                       NaN   

   positiveTestsPeopleAntibody  negativeTestsPeopleAntibody  \
0                          NaN                 

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19317 entries, 0 to 19316
Data columns (total 55 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   date                         19317 non-null  int64  
 1   state                        19317 non-null  object 
 2   positive                     19128 non-null  float64
 3   probableCases                8282 non-null   float64
 4   negative                     15379 non-null  float64
 5   pending                      2009 non-null   float64
 6   totalTestResultsSource       19317 non-null  object 
 7   totalTestResults             19215 non-null  float64
 8   hospitalizedCurrently        15960 non-null  float64
 9   hospitalizedCumulative       11903 non-null  float64
 10  inIcuCurrently               10442 non-null  float64
 11  inIcuCumulative              3480 non-null   float64
 12  onVentilatorCurrently        8290 non-null   float64
 13  onVentilatorCumu

## Metadata information

1. __checkTimeEt__<br>
_Field type:string_<br>
Deprecated<br>

2. __commercialScore__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_

3. __dataQualityGrade__<br>
Field type:string<br>
Data Quality Grade<br>
The COVID Tracking Project grade of the completeness of the data reporting by a state.<br>

4. __date__<br>
_Field type:integer_<br>
Date on which data was collected by The COVID Tracking Project.<br>

5. __dateChecked__<br>
Field type:string<br>
Deprecated<br>

6. __dateModified__<br>
_Field type:string_<br>
Deprecated, use lastUpdateEt instead<br>

7. __death__<br>
_Field type:integer_<br>
Deaths (confirmed and probable)<br>
Total fatalities with confirmed OR probable COVID-19 case diagnosis (per the expanded CSTE case definition of April 5th, 2020 approved by the CDC). In some states, these individuals must also have COVID-19 listed on the death certificate to count as a COVID-19 death. When states post multiple numbers for fatalities, the metric includes only deaths with COVID-19 listed on the death certificate, unless deaths among cases is a more reliable metric in the state.<br>
_Returns : null. if no data is available_<br>

8. __deathConfirmed__<br>
_Field type:integer_<br>
Deaths (confirmed)<br>
Total fatalities with confirmed COVID-19 case diagnosis (per the expanded CSTE case definition of April 5th, 2020 approved by the CDC). In some states, these individuals must also have COVID-19 listed on the death certificate to count as a COVID-19 death. When states post multiple numbers for confirmed fatalities, the metric includes only lab-confirmed deaths with COVID-19 listed on the death certificate, unless deaths among confirmed cases is a more reliable metric in the state.<br>
_Returns : null, if no data is available_

9. __deathIncrease__<br>
_Field type:integer_<br>
New deaths<br>
Daily increase in death, calculated from the previous day’s value.<br>
_Returns: null, if no data is available_<br>

10. __deathProbable__<br>
_Field type:integer_<br>
Deaths (probable)<br>
Total fatalities with probable COVID-19 case diagnosis (per the expanded CSTE case definition of April 5th, 2020 approved by the CDC). In some states, these individuals must also have COVID-19 listed on the death certificate to count as a COVID-19 death. When states post multiple numbers for probable fatalities, the metric includes only probable fatalities with COVID-19 listed on the death certificate, unless deaths among probable cases is a more reliable metric in the state.<br>
_Returns : null, if no data is available_<br>

11. __fips__<br>
_Field type:string_<br>
FIPS code<br>
State FIPS code<br>

12. __grade__<br>
_Field type:string_<br>
Deprecated<br>

13. __hash__<br>
_Field type:string_<br>
Deprecated - A hash of the current record.<br>

14. __hospitalized__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_

15. __hospitalizedCumulative__<br>
_Field type:integer_<br>
Cumulative hospitalized/Ever hospitalized<br>
Total number of individuals who have ever been hospitalized with COVID-19. Definitions vary by state / territory, and it is not always clear whether pediatric patients are included in this metric. Where possible, we report patients hospitalized with confirmed or suspected COVID-19 cases.<br>
_Returns : null, if no data is available_

16. __hospitalizedCurrently__<br>
_Field type:integer_<br>
Currently hospitalized/Now hospitalized<br>
Individuals who are currently hospitalized with COVID-19. Definitions vary by state / territory, and it is not always clear whether pediatric patients are included in this metric. Where possible, we report patients hospitalized with confirmed or suspected COVID-19 cases.<br>
_Returns : null, if no data is available_<br>

17. __hospitalizedIncrease__<br>
_Field type:integer_<br>
New total hospitalizations<br>
Daily increase in hospitalizedCumulative, calculated from the previous day’s value.<br>
_Returns : null, if no data is available_

18. __inIcuCumulative__<br>
_Field type:integer_<br>
Cumulative in ICU/Ever in ICU<br>
Total number of individuals who have ever been hospitalized in the Intensive Care Unit with COVID-19. Definitions vary by state / territory, and it is not always clear whether pediatric patients are included in this metric. Where possible, we report patients in the ICU with confirmed or suspected COVID-19 cases.<br>
_Returns : null, if no data is available<br>

19. __inIcuCurrently__<br>
_Field type:integer_<br>
Currently in ICU/Now in ICU<br>
Individuals who are currently hospitalized in the Intensive Care Unit with COVID-19. Definitions vary by state / territory, and it is not always clear whether pediatric patients are included in this metric. Where possible, we report patients in the ICU with confirmed or suspected COVID-19 cases.<br>
_Returns : null, if no data is available_<br>

20. __lastUpdateEt__<br>
_Field type:string_<br>
Last Update (ET)<br>
Date and time in Eastern time the state or territory last updated the data.<br>

21. __negative__<br>
_Field type:integer_<br>
Negative PCR tests (people)<br>
Total number of unique people with a completed PCR test that returns negative. For states / territories that do not report this number directly, we compute it using one of several methods, depending on which data points the state provides. Due to complex reporting procedures, this number might be mixing units and therefore, at best, it should only be considered an estimate of the number of people with a completed PCR test that return negative.<br>
_Returns : null, if no data is available_

22. __negativeIncrease__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_

23. __negativeRegularScore__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_<br>

24. __negativeScore__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_<br>

25. __negativeTestsAntibody__<br>
_Field type:integer_<br>
Negative antibody tests (specimens)<br>
The total number of completed antibody tests that return negative as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

26. __negativeTestsPeopleAntibody__<br>
_Field type:integer_<br>
Negative antibody tests (people)<br>
The total number of unique people with completed antibody tests that return negative as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

27. __negativeTestsViral__<br>
_Field type:integer_<br>
Negative PCR tests (specimens)<br>
Total number of completed PCR tests (or specimens tested) that return negative as reported by the state or territory. For states/territories that do not report this number directly, we compute it using one of several methods, depending on which data points the state provides. If we discover that a jurisdiction is including antigen tests in this metric, we will annotate that state or territory’s data accordingly.<br>
_Returns : null, if no data is available_<br>

28. __onVentilatorCumulative__<br>
_Field type:integer_<br>
Cumulative on ventilator/Ever on ventilator<br>
Total number of individuals who have ever been hospitalized under advanced ventilation with COVID-19. Definitions vary by state / territory, and it is not always clear whether pediatric patients are included in this metric. Where possible, we report patients on ventilation with confirmed or suspected COVID-19 cases.<br>
_Returns : null, if no data is available_<br>

29. __onVentilatorCurrently__<br>
_Field type:integer_<br>
Currently on ventilator/Now on ventilator<br>
Individuals who are currently hospitalized under advanced ventilation with COVID-19. Definitions vary by state / territory, and it is not always clear whether pediatric patients are included in this metric. Where possible, we report patients on ventilation with confirmed or suspected COVID-19 cases.<br>
_Returns : null, if no data is available_<br>

30. __pending__<br>
_Field type:integer_<br>
Pending<br>
Total number of viral tests that have not been completed as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

31. __posNeg__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_<br>

32. __positive__<br>
_Field type:integer_<br>
Cases (confirmed plus probable)<br>
Total number of confirmed plus probable cases of COVID-19 reported by the state or territory, ideally per the August 5, 2020 CSTE case definition. Some states are following the older April 5th, 2020 CSTE case definition or using their own custom definitions. Not all states and territories report probable cases. If a state is not reporting probable cases, this field will just represent confirmed cases.<br>
_Returns : null, if no data is available_<br>

33. __positiveCasesViral__<br>
_Field type:integer_<br>
Confirmed Cases<br>
Total number of unique people with a positive PCR or other approved nucleic acid amplification test (NAAT), as reported by the state or territory. This is equivalent to a confirmed case as per the CSTE case definitions of August 5th, 2020 and April 5th, 2020. If we discover a jurisdiction is labeling cases as confirmed using other evidence (e.g. positive antigen tests), we will annotate that state or territory’s data accordingly.<br>
_Returns : null, if no data is available_<br>

34. __positiveIncrease__<br>
_Field type:integer_<br>
New cases<br>
The daily increase in API field positive, which measures Cases (confirmed plus probable) calculated based on the previous day’s value.<br>
_Returns : null, if no data is available_<br>

35. __positiveScore__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_<br>

36. __positiveTestsAntibody__<br>
_Field type:integer_<br>
Positive antibody tests (specimens)<br>
Total number of completed antibody tests that return positive as reported by the state or territory.<br>
_Returns : null, if no data is available_

37. __positiveTestsAntigen__<br>
_Field type:integer_<br>
Positive antigen tests (specimens)<br>
Total number of completed antigen tests that return positive as reported by the state or territory.<br>
_Returns : null, if no data is available_

38. __positiveTestsPeopleAntibody__<br>
_Field type:integer_<br>
Positive antibody tests (people)<br>
The total number of unique people with completed antibody tests that return positive as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

39. __positiveTestsPeopleAntigen__<br>
_Field type:integer_<br>
Positive antigen tests (people)<br>
Total number of unique people with a completed antigen test that returned positive as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

40. __positiveTestsViral__<br>
_Field type:integer_<br>
Positive PCR tests (specimens)<br>
Total number of completed PCR tests (or specimens tested) that return positive as reported by the state or territory. If we discover that a jurisdiction is including antigen tests in this metric, we will annotate that state or territory’s data accordingly.<br>
_Returns : null, if no data is available_<br>

41. __probableCases__<br>
_Field type:integer_<br>
Probable Cases<br>
Total number of probable cases of COVID-19 as reported by the state or territory, ideally per the August 5, 2020 CSTE case definition. By this definition, a probable case is someone who tests positive via antigen without a positive PCR or other approved nucleic acid amplification test (NAAT), someone with clinical and epidemiological evidence of COVID-19 infection with no confirmatory laboratory testing performed for SARS-CoV-2, or someone with COVID-19 listed on their death certificate with no confirmatory laboratory testing performed for SARS-CoV-2. Some states are following the older April 5th, 2020 CSTE case definition or using their own custom definitions.<br>
_Returns : null, if no data is available_<br>

42. __recovered__<br>
_Field type:integer_<br>
Recovered<br>
Total number of people that are identified as recovered from COVID-19. States provide very disparate definitions on what constitutes a “recovered” COVID-19 case. Types of “recovered” cases include those who are discharged from hospitals, released from isolation after meeting CDC guidance on symptoms cessation, or those who have not been identified as fatalities after a number of days (30 or more) post disease onset. Specifics vary for each state or territory.<br>
_Returns : null, if no data is available_<br>

43. __score__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns : null, if no data is available_<br>

44. __state__<br>
_Field type:string_<br>
State (or territory)<br>
Two-letter abbreviation for the state or territory.<br>

45. __total__<br>
_Field type:integer_<br>
Deprecated<br>
_Returns :  null, if no data is available_

46. __totalTestEncountersViral__<br>
_Field type:integer_<br>
Total PCR tests (test encounters)<br>
Total number of people tested per day via PCR testing as reported by the state or territory. The count for this metric is incremented up by one for each day on which an individual person is tested, no matter how many specimens are collected from that person on that day. If an individual person is tested twice a day on three different days, this count will increment up by three. If we discover that a jurisdiction is including antigen tests in this metric, we will annotate that state or territory’s data accordingly.<br>
_Returns : null, if no data is available_<br>

47. __totalTestResults__<br>
_Field type:integer_<br>
Total test results<br>
At the national level, this metric is a summary statistic which, because of the variation in test reporting methods, is at best an estimate of US viral (PCR) testing. Some states/territories report tests in units of test encounters, some report tests in units of specimens, and some report tests in units of unique people. Moreover, some jurisdictions include antigen tests in their total test counts without reporting a separate total of viral (PCR) tests. Therefore, this value is an aggregate calculation of heterogeneous figures. Please consult each state or territory’s individual data page to see whether that jurisdiction lumps antigen and PCR tests together and to see what units that jurisdiction uses for test reporting.<br>
In most states, the totalTestResults field is currently computed by adding positive and negative values because, historically, some states do not report totals, and to work around different reporting cadences for cases and tests. In Colorado, Delaware, the District of Columbia, Florida, Hawaii, Minnesota, Nevada, New York, North Dakota, Pennsylvania, Rhode Island, Virginia, Washington, and Wisconsin, where reliable testing encounters figures are available with a complete time series, we directly report those figures in this field. In Alaska, America Samoa, Arizona, Arkansas, California, Georgia, Indiana, Kentucky, Maryland, Massachusetts, Missouri, Nebraska, New Hampshire, Ohio, Oregon, Texas, Utah, Vermont, and Wyoming, where reliable specimens figures are available with a complete time series, we directly report those figures in this field. In Alabama, Idaho, and South Dakota, where reliable unique people figures are available with a complete time series, we directly report those figures in this field. We are in the process of switching all states over to use directly reported total figures, using a policy of preferring testing encounters, specimens, and people, in that order.<br>
_Returns : null, if no data is available_<br>


48. __totalTestResultsIncrease__<br>
_Field type:integer_<br>
New tests<br>
Daily increase in totalTestResults, calculated from the previous day’s value. This calculation includes all the caveats associated with Total tests/totalTestResults, and we recommend against using it at the state/territory level.<br>
_Returns : null, if no data is available_<br>

49. __totalTestResultsSource__<br>
_Field type:string_<br>
Indicates which field is being used for total test results. If it is posNeg, then it is calculated by adding all positive and negative values.<br>

50. __totalTestsAntibody__<br>
_Field type:integer_<br>
Total antibody tests (specimens)<br>
Total number of completed antibody tests as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

51. __totalTestsAntigen__<br>
_Field type:integer_<br>
Total antigen tests (specimens)<br>
Total number of completed antigen tests, as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

52. __totalTestsPeopleAntibody__<br>
_Field type:integer_<br>
Total antibody tests (people)<br>
The total number of unique people who have been tested at least once via antibody testing as reported by the state or territory.<br>
_Returns : null, if no data is available_<br>

53. __totalTestsPeopleAntigen__<br>
_Field type:integer_<br>
Total antigen tests (people)<br>
Total number of unique people who have been tested at least once via antigen testing, as reported by the state or territory.<br>
_Returns : null, if no data is available_

54. __totalTestsPeopleViral__<br>
_Field type:integer_<br>
Total PCR tests (people)<br>
Total number of unique people tested at least once via PCR testing, as reported by the state or territory. The count for this metric is incremented up only the first time an individual person is tested and their result is reported. Future tests of the same person will not be added to this count.<br>
In the case where the state only provides negative cases, this field is calculated as the summation of people who tested positive (“Positive Cases (People”) and the number of people who tested negative (“Negative (People or Cases)”). If we discover that a jurisdiction is including antigen tests in this metric, we will annotate that state or territory’s data accordingly.<br>
_Returns : null, if no data is available_<br>


55. __totalTestsViral__<br>
_Field type:integer_<br>
Total PCR tests (specimens)<br>
Total number of PCR tests (or specimens tested) as reported by the state or territory. The count for this metric is incremented up each time a specimen is tested and the result is reported. If we discover that a jurisdiction is including antigen tests in this metric, we will annotate that state or territory’s data accordingly. For states with ambiguous annotations, we have assigned their total tests to this category; these states and territories are identified in the new API field covidTrackingProjectPreferredTotalTestUnits as having “Unclear units.”<br>
_Returns : null, if no data is available_

In [6]:
df.corr() # Checking the correlation values between the different features

date  positive  probableCases  negative  \
date                         1.000000  0.406953       0.480695  0.360615   
positive                     0.406953  1.000000       0.902621  0.860753   
probableCases                0.480695  0.902621       1.000000  0.619284   
negative                     0.360615  0.860753       0.619284  1.000000   
pending                      0.051483  0.229168       0.711549  0.255815   
totalTestResults             0.346802  0.919123       0.761658  0.915368   
hospitalizedCurrently        0.180174  0.786320       0.794900  0.609881   
hospitalizedCumulative       0.240822  0.811108       0.715896  0.855640   
inIcuCurrently               0.107367  0.737039       0.784630  0.464114   
inIcuCumulative              0.381717  0.870537       0.753781  0.854767   
onVentilatorCurrently        0.127605  0.544531       0.453590  0.321481   
onVentilatorCumulative       0.470220  0.808912       0.893374  0.913732   
recovered                    0.366914  0.868533       0.867468  0.690782   
death                        0.275478  0.848198       0.710654  0.824683   
hospitalized                 0.240822  0.811108       0.715896  0.855640   
totalTestsViral              0.336786  0.938683       0.766409  0.924448   
positiveTestsViral           0.386971  0.985458       0.905947  0.866395   
negativeTestsViral           0.365947  0.959136       0.828629  0.951608   
positiveCasesViral           0.371908  0.996960       0.874178  0.848058   
deathConfirmed               0.327119  0.776110       0.440004  0.787443   
deathProbable                0.391914  0.594548       0.515043  0.544258   
totalTestEncountersViral     0.380805  0.907358       0.862148  0.983270   
totalTestsPeopleViral        0.321356  0.954295       0.772763  0.998749   
totalTestsAntibody           0.233785  0.903696       0.737401  0.865197   
positiveTestsAntibody        0.336291  0.873832       0.773710  0.742932   
negativeTestsAntibody        0.167060  0.914100       0.706254  0.919603   
totalTestsPeopleAntibody     0.160111  0.911650       0.632823  0.938005   
positiveTestsPeopleAntibody  0.297487  0.950756       0.735426  0.938782   
negativeTestsPeopleAntibody  0.235951  0.933610       0.774399  0.972185   
totalTestsPeopleAntigen      0.720180  0.845524       0.583500  0.819421   
positiveTestsPeopleAntigen   0.826425  0.917303       0.999780  0.782108   
totalTestsAntigen            0.538442  0.792119       0.926159  0.395802   
positiveTestsAntigen         0.602670  0.806832       0.900864  0.603419   
fips                         0.001368 -0.140201      -0.078458 -0.197007   
positiveIncrease             0.230605  0.794572       0.692608  0.676981   
negativeIncrease             0.060044  0.100682       0.039888  0.291017   
total                        0.349345  0.564103       0.456224  0.998694   
totalTestResultsIncrease     0.196139  0.782284       0.503674  0.732926   
posNeg                       0.349378  0.564078       0.456138  0.998702   
deathIncrease                0.206180  0.634676       0.637974  0.449479   
hospitalizedIncrease         0.039125  0.094711       0.105635  0.105559   
commercialScore                   NaN       NaN            NaN       NaN   
negativeRegularScore              NaN       NaN            NaN       NaN   
negativeScore                     NaN       NaN            NaN       NaN   
positiveScore                     NaN       NaN            NaN       NaN   
score                             NaN       NaN            NaN       NaN   
grade                             NaN       NaN            NaN       NaN   

                              pending  totalTestResults  \
date                         0.051483          0.346802   
positive                     0.229168          0.919123   
probableCases                0.711549          0.761658   
negative                     0.255815          0.915368   
pending                      1.000000          0.224673   
totalTestResults             0

In [7]:
# Checking the correlation between the death vs all the features only
df.corr()['death']

date                           0.275478
positive                       0.848198
probableCases                  0.710654
negative                       0.824683
pending                        0.210159
totalTestResults               0.854497
hospitalizedCurrently          0.659313
hospitalizedCumulative         0.941663
inIcuCurrently                 0.601745
inIcuCumulative                0.956407
onVentilatorCurrently          0.459981
onVentilatorCumulative         0.882903
recovered                      0.563878
death                          1.000000
hospitalized                   0.941663
totalTestsViral                0.868409
positiveTestsViral             0.936315
negativeTestsViral             0.952992
positiveCasesViral             0.875308
deathConfirmed                 0.997809
deathProbable                  0.803077
totalTestEncountersViral       0.856269
totalTestsPeopleViral          0.920533
totalTestsAntibody             0.947822
positiveTestsAntibody          0.874185


In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19317 entries, 0 to 19316
Data columns (total 55 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   date                         19317 non-null  int64  
 1   state                        19317 non-null  object 
 2   positive                     19128 non-null  float64
 3   probableCases                8282 non-null   float64
 4   negative                     15379 non-null  float64
 5   pending                      2009 non-null   float64
 6   totalTestResultsSource       19317 non-null  object 
 7   totalTestResults             19215 non-null  float64
 8   hospitalizedCurrently        15960 non-null  float64
 9   hospitalizedCumulative       11903 non-null  float64
 10  inIcuCurrently               10442 non-null  float64
 11  inIcuCumulative              3480 non-null   float64
 12  onVentilatorCurrently        8290 non-null   float64
 13  onVentilatorCumu

In [9]:
# Seems the total tests done = Positive + Negative
# However there are NAN values in negative and positive columns.
# positive = 19016
# negative = 15289
# total = 19205
# The data is lagging some positive cases, many negative cases
print(df['positive'].isna().sum())
print(df['negative'].isna().sum())
print(df['total'].isna().sum())

189
3938
0


In [10]:
# Some analysis on NAN values in negative columns
df[df['negative'].isna()][['positive', 'negative', 'total']]

# It shows that, if the negative value is nan then, the total equals positive

positive  negative    total
0        53809.0       NaN    53809
5      3354591.0       NaN  3354591
8        38281.0       NaN    38281
11      947416.0       NaN   947416
13       27226.0       NaN    27226
...          ...       ...      ...
19312        0.0       NaN        0
19313        0.0       NaN        0
19314        0.0       NaN        0
19315        0.0       NaN        0
19316        NaN       NaN        0

[3938 rows x 3 columns]

In [11]:
# Some analysis on NAN values in positive columns
df[df['positive'].isna()][['positive', 'negative', 'total']]

# It shows that, if the positive value is nan then, the total equals 0

positive  negative  total
17699       NaN       NaN      0
17755       NaN       NaN      0
17811       NaN       NaN      0
17867       NaN       NaN      0
17923       NaN       NaN      0
...         ...       ...    ...
19300       NaN       NaN      0
19302       NaN       NaN      0
19304       NaN       NaN      0
19306       NaN       NaN      0
19316       NaN       NaN      0

[189 rows x 3 columns]

In [12]:
# Dropping all the columns with total = 0, which means the data is not available
df = df[df['total']!=0]

In [13]:
# Based on above filtering, checking out Nan values in positive columns
df[df['positive'].isna()][['positive', 'negative', 'total']]

positive  negative  total
18689       NaN       NaN     14
18740       NaN       NaN     14
18791       NaN       NaN      9
18842       NaN       NaN      9
18893       NaN       NaN      6
18944       NaN       NaN      2
18995       NaN       NaN      1

In [14]:
# Based on above filtering, checking out Nan values in negative columns
df[df['negative'].isna()][['positive', 'negative', 'total']]

positive  negative    total
0        53809.0       NaN    53809
5      3354591.0       NaN  3354591
8        38281.0       NaN    38281
11      947416.0       NaN   947416
13       27226.0       NaN    27226
...          ...       ...      ...
19305        2.0       NaN        2
19307        2.0       NaN        2
19308        2.0       NaN        2
19309        1.0       NaN        1
19310        1.0       NaN        1

[3691 rows x 3 columns]

In [15]:
# Looking at the 3669 scenarios, all the positives = total when negative is nan, 
#thus replacing all the nan of negative with 0
df['negative'].fillna(0, inplace=True)

In [16]:
df[df['negative'].isna()][['positive', 'negative', 'total']]

Empty DataFrame
Columns: [positive, negative, total]
Index: []

In [17]:
# Placing all the values of total to positive, when the positive is nan
df['positive'] = df.apply(lambda row: row['total'] if np.isnan(row['positive']) else row['positive'], axis=1)

In [18]:
df[df['positive'].isna()][['positive', 'negative', 'total']]

Empty DataFrame
Columns: [positive, negative, total]
Index: []

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19065 entries, 0 to 19310
Data columns (total 55 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   date                         19065 non-null  int64  
 1   state                        19065 non-null  object 
 2   positive                     19065 non-null  float64
 3   probableCases                8282 non-null   float64
 4   negative                     19065 non-null  float64
 5   pending                      2007 non-null   float64
 6   totalTestResultsSource       19065 non-null  object 
 7   totalTestResults             19004 non-null  float64
 8   hospitalizedCurrently        15960 non-null  float64
 9   hospitalizedCumulative       11899 non-null  float64
 10  inIcuCurrently               10442 non-null  float64
 11  inIcuCumulative              3480 non-null   float64
 12  onVentilatorCurrently        8290 non-null   float64
 13  onVentilatorCumu

### Lets confirm, that the model, we want should be able to predict the death cases given the total tests done, and several other relevant features
### The relevant features that should be used for prediction are as follows:

1. Date, just to divide the test and train data 
2. state, its a categorical value, need to do one hot encoding on it, because based on the state info, the results could vary.
3. positive
4. negative
5. recovered
6. hospitalized
7. hospitalizedCumulative
8. hospitalizedCurrently
9. hospitalizedIncrease
10. inIcuCurrently
11. negativeIncrease
12. onVentilatorCumulative
13. onVentilatorCurrently
14. positiveCasesViral
15. positiveIncrease


_Most of the features are dropped based on the number of non-null counts, since, 
the total number of the values in such features are very less, thus appending zeros or other values in such features
will deviate the performance and the actual data relevancy, thus will try predicting on the basis of highly available 
feature columns and the relevant feature columns only._

------
<u>Target</u> is calculating the total **deaths** based on the above features.

In [20]:
# Selecting only the above mentioned features for further processing
selected_columns = ["date", "state", "positive", "negative", "recovered", "hospitalized", 
                    "hospitalizedCumulative", "hospitalizedCurrently", "hospitalizedIncrease", 
                    "inIcuCurrently", "negativeIncrease", "onVentilatorCurrently", 
                    "positiveCasesViral", "positiveIncrease", "total", "death"]

new_df = df[selected_columns]
new_df.head()

date state  positive   negative  recovered  hospitalized  \
0  20210209    AK   53809.0        0.0        NaN        1223.0   
1  20210209    AL  474666.0  1825185.0   252880.0       43499.0   
2  20210209    AR  308848.0  2297201.0   288774.0       14174.0   
3  20210209    AS       0.0     2140.0        NaN           NaN   
4  20210209    AZ  787268.0  2834530.0   108965.0       54967.0   

   hospitalizedCumulative  hospitalizedCurrently  hospitalizedIncrease  \
0                  1223.0                   35.0                     4   
1                 43499.0                 1441.0                   116   
2                 14174.0                  775.0                    75   
3                     NaN                    NaN                     0   
4                 54967.0                 2744.0                   254   

   inIcuCurrently  negativeIncrease  onVentilatorCurrently  \
0             NaN                 0                    6.0   
1             NaN              5105                    NaN   
2           276.0              6459                  137.0   
3             NaN                 0                    NaN   
4           797.0             14216                  524.0   

   positiveCasesViral  positiveIncrease    total    death  
0                 NaN               115    53809    280.0  
1            372626.0              1318  2299851   8579.0  
2            245233.0              1475  2606049   5148.0  
3                 0.0                 0     2140      0.0  
4            735236.0              4381  3621798  14286.0

In [21]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19065 entries, 0 to 19310
Data columns (total 16 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   date                    19065 non-null  int64  
 1   state                   19065 non-null  object 
 2   positive                19065 non-null  float64
 3   negative                19065 non-null  float64
 4   recovered               13860 non-null  float64
 5   hospitalized            11899 non-null  float64
 6   hospitalizedCumulative  11899 non-null  float64
 7   hospitalizedCurrently   15960 non-null  float64
 8   hospitalizedIncrease    19065 non-null  int64  
 9   inIcuCurrently          10442 non-null  float64
 10  negativeIncrease        19065 non-null  int64  
 11  onVentilatorCurrently   8290 non-null   float64
 12  positiveCasesViral      13514 non-null  float64
 13  positiveIncrease        19065 non-null  int64  
 14  total                   19065 non-null

In [22]:
new_df.corr()['hospitalizedCumulative'] # hospitalized and hospitalizedCumulative are redundant features 

date                      0.240737
positive                  0.811134
negative                  0.394279
recovered                 0.338123
hospitalized              1.000000
hospitalizedCumulative    1.000000
hospitalizedCurrently     0.624318
hospitalizedIncrease      0.179892
inIcuCurrently            0.478860
negativeIncrease          0.059492
onVentilatorCurrently     0.555756
positiveCasesViral        0.918554
positiveIncrease          0.585150
total                     0.474030
death                     0.941655
Name: hospitalizedCumulative, dtype: float64

In [25]:
# dropping redundant features
new_df.drop('hospitalizedCumulative', inplace=True, axis=1)

/Users/adarsh/Desktop/Machine Learning/Projects/Assignment 1 Covid Dataset/.venv/lib/python3.8/site-packages/pandas/core/frame.py:4305: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [23]:
new_df.corr()

date  positive  negative  recovered  hospitalized  \
date                    1.000000  0.406419  0.305526   0.366914      0.240737   
positive                0.406419  1.000000  0.436053   0.868533      0.811134   
negative                0.305526  0.436053  1.000000   0.354185      0.394279   
recovered               0.366914  0.868533  0.354185   1.000000      0.338123   
hospitalized            0.240737  0.811134  0.394279   0.338123      1.000000   
hospitalizedCumulative  0.240737  0.811134  0.394279   0.338123      1.000000   
hospitalizedCurrently   0.180174  0.786320  0.203509   0.607317      0.624318   
hospitalizedIncrease    0.037809  0.094423  0.075853   0.029930      0.179892   
inIcuCurrently          0.107367  0.737039  0.071128   0.593645      0.478860   
negativeIncrease        0.058875  0.100421  0.298026   0.147334      0.059492   
onVentilatorCurrently   0.127605  0.544531  0.244131   0.372692      0.555756   
positiveCasesViral      0.371514  0.996960  0.365642   0.921879      0.918554   
positiveIncrease        0.228162  0.794420  0.294047   0.660173      0.585150   
total                   0.346821  0.563701  0.989119   0.462888      0.474030   
death                   0.274844  0.848191  0.454716   0.563878      0.941655   

                        hospitalizedCumulative  hospitalizedCurrently  \
date                                  0.240737               0.180174   
positive                              0.811134               0.786320   
negative                              0.394279               0.203509   
recovered                             0.338123               0.607317   
hospitalized                          1.000000               0.624318   
hospitalizedCumulative                1.000000               0.624318   
hospitalizedCurrently                 0.624318               1.000000   
hospitalizedIncrease                  0.179892               0.211992   
inIcuCurrently                        0.478860               0.979109   
negativeIncrease                      0.059492               0.051888   
onVentilatorCurrently                 0.555756               0.912690   
positiveCasesViral                    0.918554               0.852393   
positiveIncrease                      0.585150               0.840262   
total                                 0.474030               0.316689   
death                                 0.941655               0.659313   

                        hospitalizedIncrease  inIcuCurrently  \
date                                0.037809        0.107367   
positive                            0.094423        0.737039   
negative                            0.075853        0.071128   
recovered                           0.029930        0.593645   
hospitalized                        0.179892        0.478860   
hospitalizedCumulative              0.179892        0.478860   
hospitalizedCurrently               0.211992        0.979109   
hospitalizedIncrease                1.000000        0.204450   
inIcuCurrently                      0.204450        1.000000   
negativeIncrease                    0.018850        0.015777   
onVentilatorCurrently               0.128417        0.971707   
positiveCasesViral                  0.095962        0.840572   
positiveIncrease                    0.140390        0.765410   
total                               0.085084        0.191380   
death                               0.102985        0.601745   

                        negativeIncrease  onVentilatorCurrently  \
date                            0.058875               0.127605   
positive                        0.100421               0.544531   
negative                        0.298026               0.244131   
recovered                       0.147334               0.372692   
hospitalized                    0.059492               0.555756   
hospitalizedCumulative          0.059492               0.555756   
hospitalizedCurrently           0.051888               0.912690   
h

In [24]:
new_df.head()

date state  positive   negative  recovered  hospitalized  \
0  20210209    AK   53809.0        0.0        NaN        1223.0   
1  20210209    AL  474666.0  1825185.0   252880.0       43499.0   
2  20210209    AR  308848.0  2297201.0   288774.0       14174.0   
3  20210209    AS       0.0     2140.0        NaN           NaN   
4  20210209    AZ  787268.0  2834530.0   108965.0       54967.0   

   hospitalizedCumulative  hospitalizedCurrently  hospitalizedIncrease  \
0                  1223.0                   35.0                     4   
1                 43499.0                 1441.0                   116   
2                 14174.0                  775.0                    75   
3                     NaN                    NaN                     0   
4                 54967.0                 2744.0                   254   

   inIcuCurrently  negativeIncrease  onVentilatorCurrently  \
0             NaN                 0                    6.0   
1             NaN              5105                    NaN   
2           276.0              6459                  137.0   
3             NaN                 0                    NaN   
4           797.0             14216                  524.0   

   positiveCasesViral  positiveIncrease    total    death  
0                 NaN               115    53809    280.0  
1            372626.0              1318  2299851   8579.0  
2            245233.0              1475  2606049   5148.0  
3                 0.0                 0     2140      0.0  
4            735236.0              4381  3621798  14286.0

In [25]:
new_df['date'] =  pd.to_datetime(new_df['date'], format='%Y%m%d')

<ipython-input-25-214f810a4fa9>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['date'] =  pd.to_datetime(new_df['date'], format='%Y%m%d')


In [26]:
new_df.head()

date state  positive   negative  recovered  hospitalized  \
0 2021-02-09    AK   53809.0        0.0        NaN        1223.0   
1 2021-02-09    AL  474666.0  1825185.0   252880.0       43499.0   
2 2021-02-09    AR  308848.0  2297201.0   288774.0       14174.0   
3 2021-02-09    AS       0.0     2140.0        NaN           NaN   
4 2021-02-09    AZ  787268.0  2834530.0   108965.0       54967.0   

   hospitalizedCumulative  hospitalizedCurrently  hospitalizedIncrease  \
0                  1223.0                   35.0                     4   
1                 43499.0                 1441.0                   116   
2                 14174.0                  775.0                    75   
3                     NaN                    NaN                     0   
4                 54967.0                 2744.0                   254   

   inIcuCurrently  negativeIncrease  onVentilatorCurrently  \
0             NaN                 0                    6.0   
1             NaN              5105                    NaN   
2           276.0              6459                  137.0   
3             NaN                 0                    NaN   
4           797.0             14216                  524.0   

   positiveCasesViral  positiveIncrease    total    death  
0                 NaN               115    53809    280.0  
1            372626.0              1318  2299851   8579.0  
2            245233.0              1475  2606049   5148.0  
3                 0.0                 0     2140      0.0  
4            735236.0              4381  3621798  14286.0

In [27]:
new_df["state"].unique()

array(['AK', 'AL', 'AR', 'AS', 'AZ', 'CA', 'CO', 'CT', 'DC', 'DE', 'FL',
       'GA', 'GU', 'HI', 'IA', 'ID', 'IL', 'IN', 'KS', 'KY', 'LA', 'MA',
       'MD', 'ME', 'MI', 'MN', 'MO', 'MP', 'MS', 'MT', 'NC', 'ND', 'NE',
       'NH', 'NJ', 'NM', 'NV', 'NY', 'OH', 'OK', 'OR', 'PA', 'PR', 'RI',
       'SC', 'SD', 'TN', 'TX', 'UT', 'VA', 'VI', 'VT', 'WA', 'WI', 'WV',
       'WY'], dtype=object)

### One hot encoding the states column

In [28]:
from sklearn.preprocessing import OneHotEncoder
# creating instance of one-hot-encoder
enc = OneHotEncoder(handle_unknown='ignore')
# passing state column (label encoded values of states)
enc_df = pd.DataFrame(enc.fit_transform(new_df[['state']]).toarray())
# merge with new_df on key values
new_df = new_df.join(enc_df)
new_df.head()
# enc_df

date state  positive   negative  recovered  hospitalized  \
0 2021-02-09    AK   53809.0        0.0        NaN        1223.0   
1 2021-02-09    AL  474666.0  1825185.0   252880.0       43499.0   
2 2021-02-09    AR  308848.0  2297201.0   288774.0       14174.0   
3 2021-02-09    AS       0.0     2140.0        NaN           NaN   
4 2021-02-09    AZ  787268.0  2834530.0   108965.0       54967.0   

   hospitalizedCumulative  hospitalizedCurrently  hospitalizedIncrease  \
0                  1223.0                   35.0                     4   
1                 43499.0                 1441.0                   116   
2                 14174.0                  775.0                    75   
3                     NaN                    NaN                     0   
4                 54967.0                 2744.0                   254   

   inIcuCurrently  negativeIncrease  onVentilatorCurrently  \
0             NaN                 0                    6.0   
1             NaN              5105                    NaN   
2           276.0              6459                  137.0   
3             NaN                 0                    NaN   
4           797.0             14216                  524.0   

   positiveCasesViral  positiveIncrease    total    death    0    1    2    3  \
0                 NaN               115    53809    280.0  1.0  0.0  0.0  0.0   
1            372626.0              1318  2299851   8579.0  0.0  1.0  0.0  0.0   
2            245233.0              1475  2606049   5148.0  0.0  0.0  1.0  0.0   
3                 0.0                 0     2140      0.0  0.0  0.0  0.0  1.0   
4            735236.0              4381  3621798  14286.0  0.0  0.0  0.0  0.0   

     4    5    6    7    8    9   10   11   12   13   14   15   16   17   18  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    19   20   21   22   23   24   25   26   27   28   29   30   31   32   33  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    34   35   36   37   38   39   40   41   42   43   44   45   46   47   48  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    49   50   51   52   53   54   55  
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0

In [29]:
# Lets drop the state column now, since its already one hot encoded
new_df.drop('state', inplace=True, axis=1)

In [30]:
new_df.head()

date  positive   negative  recovered  hospitalized  \
0 2021-02-09   53809.0        0.0        NaN        1223.0   
1 2021-02-09  474666.0  1825185.0   252880.0       43499.0   
2 2021-02-09  308848.0  2297201.0   288774.0       14174.0   
3 2021-02-09       0.0     2140.0        NaN           NaN   
4 2021-02-09  787268.0  2834530.0   108965.0       54967.0   

   hospitalizedCumulative  hospitalizedCurrently  hospitalizedIncrease  \
0                  1223.0                   35.0                     4   
1                 43499.0                 1441.0                   116   
2                 14174.0                  775.0                    75   
3                     NaN                    NaN                     0   
4                 54967.0                 2744.0                   254   

   inIcuCurrently  negativeIncrease  onVentilatorCurrently  \
0             NaN                 0                    6.0   
1             NaN              5105                    NaN   
2           276.0              6459                  137.0   
3             NaN                 0                    NaN   
4           797.0             14216                  524.0   

   positiveCasesViral  positiveIncrease    total    death    0    1    2    3  \
0                 NaN               115    53809    280.0  1.0  0.0  0.0  0.0   
1            372626.0              1318  2299851   8579.0  0.0  1.0  0.0  0.0   
2            245233.0              1475  2606049   5148.0  0.0  0.0  1.0  0.0   
3                 0.0                 0     2140      0.0  0.0  0.0  0.0  1.0   
4            735236.0              4381  3621798  14286.0  0.0  0.0  0.0  0.0   

     4    5    6    7    8    9   10   11   12   13   14   15   16   17   18  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    19   20   21   22   23   24   25   26   27   28   29   30   31   32   33  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    34   35   36   37   38   39   40   41   42   43   44   45   46   47   48  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    49   50   51   52   53   54   55  
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0

In [32]:
# Counting NAN in death column
new_df['death'].isna().sum()

670

In [33]:
# Dropping rows with death feature nan
new_df = new_df[new_df['death'].notna()]
new_df.head()

date  positive   negative  recovered  hospitalized  \
0 2021-02-09   53809.0        0.0        NaN        1223.0   
1 2021-02-09  474666.0  1825185.0   252880.0       43499.0   
2 2021-02-09  308848.0  2297201.0   288774.0       14174.0   
3 2021-02-09       0.0     2140.0        NaN           NaN   
4 2021-02-09  787268.0  2834530.0   108965.0       54967.0   

   hospitalizedCumulative  hospitalizedCurrently  hospitalizedIncrease  \
0                  1223.0                   35.0                     4   
1                 43499.0                 1441.0                   116   
2                 14174.0                  775.0                    75   
3                     NaN                    NaN                     0   
4                 54967.0                 2744.0                   254   

   inIcuCurrently  negativeIncrease  onVentilatorCurrently  \
0             NaN                 0                    6.0   
1             NaN              5105                    NaN   
2           276.0              6459                  137.0   
3             NaN                 0                    NaN   
4           797.0             14216                  524.0   

   positiveCasesViral  positiveIncrease    total    death    0    1    2    3  \
0                 NaN               115    53809    280.0  1.0  0.0  0.0  0.0   
1            372626.0              1318  2299851   8579.0  0.0  1.0  0.0  0.0   
2            245233.0              1475  2606049   5148.0  0.0  0.0  1.0  0.0   
3                 0.0                 0     2140      0.0  0.0  0.0  0.0  1.0   
4            735236.0              4381  3621798  14286.0  0.0  0.0  0.0  0.0   

     4    5    6    7    8    9   10   11   12   13   14   15   16   17   18  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    19   20   21   22   23   24   25   26   27   28   29   30   31   32   33  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    34   35   36   37   38   39   40   41   42   43   44   45   46   47   48  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    49   50   51   52   53   54   55  
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0

In [34]:
# Counting NAN in death column
new_df['death'].isna().sum()

0

In [35]:
# Lets fill all the Nans with zeros
new_df.fillna(0, inplace=True)

In [36]:
new_df.head()

date  positive   negative  recovered  hospitalized  \
0 2021-02-09   53809.0        0.0        0.0        1223.0   
1 2021-02-09  474666.0  1825185.0   252880.0       43499.0   
2 2021-02-09  308848.0  2297201.0   288774.0       14174.0   
3 2021-02-09       0.0     2140.0        0.0           0.0   
4 2021-02-09  787268.0  2834530.0   108965.0       54967.0   

   hospitalizedCumulative  hospitalizedCurrently  hospitalizedIncrease  \
0                  1223.0                   35.0                     4   
1                 43499.0                 1441.0                   116   
2                 14174.0                  775.0                    75   
3                     0.0                    0.0                     0   
4                 54967.0                 2744.0                   254   

   inIcuCurrently  negativeIncrease  onVentilatorCurrently  \
0             0.0                 0                    6.0   
1             0.0              5105                    0.0   
2           276.0              6459                  137.0   
3             0.0                 0                    0.0   
4           797.0             14216                  524.0   

   positiveCasesViral  positiveIncrease    total    death    0    1    2    3  \
0                 0.0               115    53809    280.0  1.0  0.0  0.0  0.0   
1            372626.0              1318  2299851   8579.0  0.0  1.0  0.0  0.0   
2            245233.0              1475  2606049   5148.0  0.0  0.0  1.0  0.0   
3                 0.0                 0     2140      0.0  0.0  0.0  0.0  1.0   
4            735236.0              4381  3621798  14286.0  0.0  0.0  0.0  0.0   

     4    5    6    7    8    9   10   11   12   13   14   15   16   17   18  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    19   20   21   22   23   24   25   26   27   28   29   30   31   32   33  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    34   35   36   37   38   39   40   41   42   43   44   45   46   47   48  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

    49   50   51   52   53   54   55  
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0

### Dividing data into train, and test sets

In [37]:
# Since the dataset is ordered in chronological manner, 
# lets put 70% of past data as training data and latest 30% as test data
mask = round(0.3*len(new_df))
print(mask)
test = new_df[:mask]
train = new_df[mask:]

5518


In [38]:
len(train)

12877

In [39]:
len(test)

5518

In [40]:
train.drop("date", inplace=True, axis=1)
test.drop('date', inplace=True, axis=1)

/Users/adarsh/Desktop/Machine Learning/Projects/Assignment 1 Covid Dataset/.venv/lib/python3.8/site-packages/pandas/core/frame.py:4305: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [41]:
# resetting the index 
train.reset_index(inplace=True, drop=True)

In [42]:
train_features = list(set(train.columns) - set(['death']))
# print(train_features)
X_train = train[train_features]
y_train = train["death"]
X_test = test[train_features]
y_test = test["death"]

### Simple Linear regression without any feature manipulations

In [43]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [44]:
simple_regressor = LinearRegression()
simple_regressor.fit(X_train, y_train)

LinearRegression()

In [45]:
predictions = simple_regressor.predict(X_train)

In [46]:
print('RMSE for Simple Linear Regression : ',np.sqrt(mean_squared_error(y_train, predictions)))

RMSE for Simple Linear Regression :  718.8795566184416


#### Standardization of feature columns

In [47]:
from sklearn.preprocessing import StandardScaler

In [48]:
scaler = StandardScaler()

In [49]:
scaler.fit(X_train)

StandardScaler()

In [50]:
scaler.transform(X_train)

array([[-0.13425774, -0.13515415, -0.13485596, ..., -0.37622964,
         1.85599261, -0.14596869],
       [-0.13425774, -0.13515415, -0.13485596, ..., -0.23720744,
        -0.06542445,  0.03756758],
       [-0.13425774, -0.13515415, -0.13485596, ..., -0.12497682,
         0.15587616, -0.0422308 ],
       ...,
       [-0.13425774, -0.13515415, -0.13485596, ..., -0.37622964,
        -0.47392671, -0.14596869],
       [-0.13425774, -0.13515415, -0.13485596, ..., -0.37622964,
        -0.47394406, -0.14596869],
       [-0.13425774, -0.13515415, -0.13485596, ..., -0.37622964,
        -0.47395273, -0.14596869]])

In [51]:
standardized_regressor = LinearRegression()
standardized_regressor.fit(scaler.transform(X_train), y_train)

LinearRegression()

In [52]:
standardized_predictions = standardized_regressor.predict(scaler.transform(X_train))

In [53]:
print('RMSE for Standardized Linear Regression : ',np.sqrt(mean_squared_error(y_train, standardized_predictions)))

RMSE for Standardized Linear Regression :  718.6090993732389


Clearly, the above result of RMSE is pretty bad.

-------

#### Trying out polynomial features for fitting the data

In [55]:
from sklearn.preprocessing import PolynomialFeatures

#### Polynomial regression with degree 2

In [56]:
poly_2 = PolynomialFeatures(degree = 2) 
X_poly_2 = poly_2.fit_transform(scaler.transform(X_train)) # First scaling to standard form, them implementing polynomial feature transformation

# poly_scaler_2 = StandardScaler()
# poly_scaler_2.fit(X_poly_2)

poly_reg_2 = LinearRegression() 
poly_reg_2.fit(X_poly_2, y_train) 
pred_2 = poly_reg_2.predict(X_poly_2)

print('RMSE for polynomial features with degree 2 Regression :', np.sqrt(mean_squared_error(y_train, pred_2)))

RMSE for polynomial features with degree 2 Regression : 82.84222031364514


Degree 2 for the polynomial feature has reduced the RMSE value by almost 10 times, the model is obviously not underfitting. 

----

#### Polynomial regression with degree 3

In [57]:
poly_3 = PolynomialFeatures(degree = 3) 
X_poly_3 = poly_3.fit_transform(scaler.transform(X_train))

# poly_scaler_3 = StandardScaler()
# poly_scaler_3.fit(X_poly_3)

poly_reg_3 = LinearRegression() 
poly_reg_3.fit(X_poly_3, y_train) 
pred_3 = poly_reg_3.predict(X_poly_3)

print('RMSE for polynomial features with degree 3 Regression :', np.sqrt(mean_squared_error(y_train, pred_3)))

RMSE for polynomial features with degree 3 Regression : 17.101975419608642


### Result analysis on test set

The linear regression with polynomial feature has decreased the RMSE error significantly.
The degree 3 polynomial feature has the lowest RMSE in training data, that is 17.1.

In [59]:
# Lets check the model polynomial regressionn model performance on test set
print("RMSE for test data with polynomial regression of degree 3 : ", np.sqrt(mean_squared_error(y_test, 
                                                                                                 poly_reg_3.predict(poly_3.transform(scaler.transform(X_test))))))

RMSE for test data with polynomial regression of degree 3 :  8185601370.174896


It can be clearly seen from the RMSE score that the polynomial regression model of degree 3 is overfitting alot

In [60]:
# Lets check if the model is overfitting in polynomial regression with degree 2.
test_poly_2 = PolynomialFeatures(degree = 2) 
X_test_2 = test_poly_2.fit_transform(X_test)
print("RMSE for test data with polynomial regression of degree 2 : ", np.sqrt(mean_squared_error(y_test, 
                                                                                                 poly_reg_2.predict(poly_2.transform(scaler.transform(X_test))))))

RMSE for test data with polynomial regression of degree 2 :  782473173696.663


Polynomial regression model of degree 2 is also very overfitting

In [61]:
print("RMSE for test data with standardized regressor : ", np.sqrt(mean_squared_error(y_test, standardized_regressor.predict(scaler.transform(X_test)))))

RMSE for test data with standardized regressor :  3052.4408211501514


------

### Applying regularization

#### L2 regularization

#### L2 regularization on linear regression

In [62]:
from sklearn.linear_model import Ridge
ridge_standardized = Ridge(alpha=100.0)

In [63]:
ridge_standardized.fit(scaler.transform(X_train), y_train)

Ridge(alpha=100.0)

In [64]:
print('RMSE on training data for Standardized Ridge Regression : ',np.sqrt(mean_squared_error(y_train, ridge_standardized.predict(scaler.transform(X_train)))))

RMSE on training data for Standardized Ridge Regression :  729.1063125940465


In [65]:
print("RMSE for test data with ridge regressor : ", np.sqrt(mean_squared_error(y_test, ridge_standardized.predict(scaler.transform(X_test)))))

RMSE for test data with ridge regressor :  2869.1537353568906


As from the result of regularization on standard linear regression, it can be seen that the model performance on test data has improved slightly as compared to the without regularization result.

#### L2 regularization on polynomial regression of degree 2

In [66]:
ridge_poly_2 = Ridge(alpha=100.0)

In [67]:
ridge_poly_2.fit(poly_2.transform(scaler.transform(X_train)), y_train)

Ridge(alpha=100.0)

In [68]:
print('RMSE on training data for Standardized Ridge Regression with degree 2 : ',np.sqrt(mean_squared_error(y_train, ridge_poly_2.predict(poly_2.transform(scaler.transform(X_train))))))

RMSE on training data for Standardized Ridge Regression with degree 2 :  147.33576975945365


In [69]:
print('RMSE on test data for Standardized Ridge Regression with degree 2 : ',np.sqrt(mean_squared_error(y_test, ridge_poly_2.predict(poly_2.transform(scaler.transform(X_test))))))

RMSE on test data for Standardized Ridge Regression with degree 2 :  11026.282836018423


By adding L2 regularization on polynomial regression of order 2, the model performance on test data improved by factor of 10^12 to 10^4. The addition of L2 regularization on the model hada huge improvement interms of the model generalizability at a cost of some increased trainig error.

#### L2 regularization on polynomial regression of degree 3

In [70]:
ridge_poly_3 = Ridge(alpha=100.0)

In [71]:
ridge_poly_3.fit(poly_3.transform(scaler.transform(X_train)), y_train)

Ridge(alpha=100.0)

In [170]:
print('RMSE on training data for Standardized Ridge Regression with degree 3 : ',np.sqrt(mean_squared_error(y_train, ridge_poly_3.predict(poly_3.transform(scaler.transform(X_train))))))

RMSE on training data for Standardized Ridge Regression with degree 3 :  28.74896331312586


In [171]:
print('RMSE on test data for Standardized Ridge Regression with degree 3 : ',np.sqrt(mean_squared_error(y_test, ridge_poly_3.predict(poly_scaler_3.transform(poly_3.transform(X_test))))))

RMSE on test data for Standardized Ridge Regression with degree 3 :  17804.929700532004


By adding L2 regularization on polynomial regression of order 3, the model performance on test data improved by 7.6\*10^16 to 17804. The generalizability of the model has improved very highly. From the result of normal polynomial regression with degree 3 the training error was only 17 however the test error was in 10^16 which is a very high error, and that was showing that the model was overfitting. But by adding regularization the training error increased by a little however the test error reduced from factor of 10^16 to 10^3.
Eventhough, the model performance is not good, but the adding of penalty and its effect on model generalizibility can be clearly seen. 

------

#### L1 regularization

#### L1 regularization on linear regression

In [180]:
from sklearn.linear_model import Lasso
lasso_standardized = Lasso(alpha=100.0)

In [181]:
lasso_standardized.fit(scaler.transform(X_train), y_train)

Lasso(alpha=100.0)

In [182]:
print('RMSE on training data for Standardized lasso Regression : ',np.sqrt(mean_squared_error(y_train, lasso_standardized.predict(scaler.transform(X_train)))))

RMSE on training data for Standardized Ridge Regression :  953.3144018862494


In [183]:
print("RMSE for test data with lasso regressor : ", np.sqrt(mean_squared_error(y_test, lasso_standardized.predict(scaler.transform(X_test)))))

RMSE for test data with ridge regressor :  3070.182596553438


As from the result of regularization on standard linear regression, it can be seen that the model performance on test data has improved slightly as compared to the without regularization result. In addition, the model performance with L1 or L2 regression are quiet comparable

#### L1 regularization on polynomial regression of degree 2

In [184]:
lasso_poly_2 = Lasso(alpha=100.0)

In [185]:
lasso_poly_2.fit(poly_scaler_2.transform(poly_2.transform(X_train)), y_train)

Lasso(alpha=100.0)

In [186]:
print('RMSE on training data for Standardized Lasso Regression with degree 2 : ',np.sqrt(mean_squared_error(y_train, lasso_poly_2.predict(poly_scaler_2.transform(poly_2.transform(X_train))))))

RMSE on training data for Standardized Lasso Regression with degree 2 :  607.7821862627763


In [187]:
print('RMSE on test data for Standardized Lasso Regression with degree 2 : ',np.sqrt(mean_squared_error(y_test, lasso_poly_2.predict(poly_scaler_2.transform(poly_2.transform(X_test))))))

RMSE on test data for Standardized Lasso Regression with degree 2 :  4599.883232362514


By adding L1 regularization on polynomial regression of order 2, the model performance on test data improved by 1.06\*10^15 to 4599. The addition of L1 regularization on the model has scaled the test error from factor of 10^15 to 1K, which is a huge improvement interms of the model generalizability.

#### L1 regularization on polynomial regression of degree 3

In [188]:
lasso_poly_3 = Lasso(alpha=100.0)

In [189]:
lasso_poly_3.fit(poly_scaler_3.transform(poly_3.transform(X_train)), y_train)

Lasso(alpha=100.0)

In [190]:
print('RMSE on training data for Standardized Lasso Regression with degree 3 : ',np.sqrt(mean_squared_error(y_train, lasso_poly_3.predict(poly_scaler_3.transform(poly_3.transform(X_train))))))

RMSE on training data for Standardized Lasso Regression with degree 3 :  598.5867956501127


In [191]:
print('RMSE on test data for Standardized Lasso Regression with degree 3 : ',np.sqrt(mean_squared_error(y_test, lasso_poly_3.predict(poly_scaler_3.transform(poly_3.transform(X_test))))))

RMSE on test data for Standardized Lasso Regression with degree 3 :  5682.515308429499


By adding L1 regularization on polynomial regression of order 3, the model performance on test data improved by 7.6\*10^16 to 5682. The generalizability of the model has improved very highly. From the result of normal polynomial regression with degree 3 the training error was only 17 however the test error was in 10^16 which is a very high error, and that was showing that the model was overfitting. But by adding regularization the training error increased by a little however the test error reduced from factor of 10^16 to 10^3. The L1 regularization has shown better generalizibility than the L2 regularization for the polynomial regression of order 3.
Eventhough, the model performance is not good, but the adding of penalty and its effect on model generalizibility can be clearly seen. 

------

#### L1 and L2 regularization

#### L1/L2 regularization on linear regression

In [192]:
from sklearn.linear_model import ElasticNet

In [200]:
elastic_standardized = ElasticNet(alpha=1.0)

In [201]:
elastic_standardized.fit(scaler.transform(X_train), y_train)

ElasticNet()

In [202]:
print('RMSE on training data for Standardized lasso Regression : ',np.sqrt(mean_squared_error(y_train, elastic_standardized.predict(scaler.transform(X_train)))))

RMSE on training data for Standardized lasso Regression :  1191.3545334885946


In [203]:
print("RMSE for test data with lasso regressor : ", np.sqrt(mean_squared_error(y_test, elastic_standardized.predict(scaler.transform(X_test)))))

RMSE for test data with lasso regressor :  2419.857338129787


As from the result of regularization on standard linear regression, it can be seen that the model performance on test data has improved slightly as compared to the without regularization result. In addition, the model performance with L1, L2, regularization are quiet comparable. However, the L1/L2 regularization with very less penalty has very good performance.

#### L1/L2 regularization on polynomial regression of degree 2

In [212]:
elastic_poly_2 = ElasticNet(alpha=1.0)

In [213]:
elastic_poly_2.fit(poly_scaler_2.transform(poly_2.transform(X_train)), y_train)

ElasticNet()

In [216]:
print('RMSE on training data for Standardized Elastic Regression with degree 2 : ',np.sqrt(mean_squared_error(y_train, elastic_poly_2.predict(poly_scaler_2.transform(poly_2.transform(X_train))))))

RMSE on training data for Standardized Elastic Regression with degree 2 :  498.7737480143985


In [217]:
print('RMSE on test data for Standardized Elastic Regression with degree 2 : ',np.sqrt(mean_squared_error(y_test, elastic_poly_2.predict(poly_scaler_2.transform(poly_2.transform(X_test))))))

RMSE on test data for Standardized Elastic Regression with degree 2 :  9373.648415816955


By adding L1/L2 regularization on polynomial regression of order 2, the model performance on test data improved by 1.06\*10^15 to 9373. The addition of L1/L2 regularization on the model has scaled the test error from factor of 10^15 to 10K, which is a huge improvement interms of the model generalizability. However, the model performance is not par as compared to the L1 only and L2 only regularization.

#### L1/L2 regularization on polynomial regression of degree 3

In [218]:
elastic_poly_3 = ElasticNet(alpha=1.0)

In [219]:
elastic_poly_3.fit(poly_scaler_3.transform(poly_3.transform(X_train)), y_train)

/Users/adarsh/Desktop/Machine Learning/Projects/Assignment 1 Covid Dataset/.venv/lib/python3.8/site-packages/sklearn/linear_model/_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 2448670537.5238323, tolerance: 22177518.87666834
  model = cd_fast.enet_coordinate_descent(


ElasticNet()

In [220]:
print('RMSE on training data for Standardized Elastic Regression with degree 3 : ',np.sqrt(mean_squared_error(y_train, elastic_poly_3.predict(poly_scaler_3.transform(poly_3.transform(X_train))))))

RMSE on training data for Standardized Elastic Regression with degree 3 :  346.99643131125737


In [221]:
print('RMSE on test data for Standardized Elastic Regression with degree 3 : ',np.sqrt(mean_squared_error(y_test, elastic_poly_3.predict(poly_scaler_3.transform(poly_3.transform(X_test))))))

RMSE on test data for Standardized Elastic Regression with degree 3 :  18681.878235384993


By adding L1/L2 regularization on polynomial regression of order 3, the model performance on test data improved by 7.6\*10^16 to 18681. The generalizability of the model has improved very highly. From the result of normal polynomial regression with degree 3 the training error was only 17 however the test error was in 10^16 which is a very high error, and that was showing that the model was overfitting. But by adding regularization the training error increased by a little however the test error reduced from factor of 10^16 to 10^4.
Eventhough, the model performance is not good, but the adding of penalty and its effect on model generalizibility can be clearly seen. 

##### Lastly, the addition of the regularization has improved the model generalazibility, plus improved the training time by alot. However, the model performance is not as expected, but impact of regularization can be clearly seen from the results above. The reason behind the bad model performance is because of the data. The NAN value filled with zeros is one of the reason why the model was not able to perform well. Since, introducing zeros introduces a noise in the data, which means the data behavior is completely changed due to it. Secondly, the model was not performing good enough because the polynomial features that has been introduced as also not able to capture the trend and relation in data, thus a deeper data analysis is required such that more significant and more meaningful features can be taken into consideration for the development of the model. 